### PR TITLE
End-to-end partition prunability tests [HZ-2787]

### PR DESCRIPTION
**Pull request by Fly-Style**
Also refactor and optimize `PlanExecutor` a bit

### CODE REVIEW

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/SqlPartitionPruningE2ETest.java [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+                sql.getSchema(),
+                query,
+                sql.getParameters(),
+                SqlExpectedResultType.ROWS
+        );
+
+        assertInstanceOf(SqlPlanImpl.SelectPlan.class, plan);
+        SqlPlanImpl.SelectPlan selectPlan = (SqlPlanImpl.SelectPlan) plan;
+
+        var optResult = planExecutor.tryUsePrunability(selectPlan, EEC);
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/SqlPartitionPruningE2ETest.java:106]: this definitely tests more that before (it is possible now to write assertions on the DAG), but it still does not test all assumptions on the Jet-SQL interface, in particular correctness of the `JobConfig` (we might create it in a wrong way, forget to invoke tryUsePrunability in production code etc.)

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java [hunk 2]
```diff
@@ -517,64 +518,17 @@ SqlResult execute(SelectPlan plan, QueryId queryId, List<Object> arguments, long
...
-                final Partition partition = hazelcastInstance.getPartitionService().getPartition(
-                        PartitioningStrategyUtil.constructAttributeBasedKey(partitionKeyComponents)
-                );
-                if (partition == null) {
-                    // Can happen if the cluster is mid-repartitioning/migration, in this case we revert to
-                    // non-pruning logic. Alternative scenario is if the produced partitioning key somehow invalid.
-                    allVariantsValid = false;
-                    break;
+            if (!partitions.isEmpty() && allVariantsValid) {
+                if (plan.requiredRootPartitionId() != null) {
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java:527]: you will have conflicts here #25062

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java [hunk 3]
```diff
@@ -804,6 +758,62 @@ SqlResult execute(CreateTypePlan plan) {
...
+                if (partition == null) {
+                    // Can happen if the cluster is mid-repartitioning/migration, in this case we revert to
+                    // non-pruning logic. Alternative scenario is if the produced partitioning key somehow invalid.
+                    allVariantsValid = false;
+                    break;
+                }
+                partitions.add(partition.getPartitionId());
+            }
+        }
+        return Tuple2.tuple2(partitions, allVariantsValid);
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java:814]: can't we just return something empty when `!allVariantsValid`?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java (hunk 2 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java:527]: Ofc I will update this PR after the #25062  merge.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlUnionTest.java [hunk 4]
```diff
@@ -259,126 +257,6 @@ public void watermarkedStreamUnionAllTest() {
                 5);
     }
 
-    @Test
```
[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlUnionTest.java:262]: Yes. Actually, existing tests already duplicating them, so, we can remove them safely.

[the remark this answers is not in the document]
Comment (k-jamroz) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlUnionTest.java:262]: SqlPartitionPruningE2ETest has much more boilerplate so it is less readable as a test proving that we get correct results, which is more important than ensuring that we use partition pruning. SqlPartitionPruningE2ETest also executes the query in stages, there is no guarantee that code in test is exactly the same sequence as in normal execution.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java (hunk 3 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java]: Refactored  78d4a04336c63b2a30eefc3aafba35ce24c282df

**Comment by Fly-Style** [comment]
run-lab-run

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java [hunk 5]
```diff
@@ -579,6 +531,7 @@ SqlResult execute(SelectPlan plan, QueryId queryId, List<Object> arguments, long
         Object oldValue = resultRegistry.store(jobId, queryResultProducer);
         assert oldValue == null : oldValue;
         try {
+            preJobInvocationObservers.forEach(observer -> observer.onJobInvocation(plan.getDag(), jobConfig));
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java:534]: `DmlPlan` also creates a job

File: hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+package com.hazelcast.jet;
+
+import com.hazelcast.internal.cluster.MemberInfo;
+import com.hazelcast.jet.config.JobConfig;
+import com.hazelcast.jet.core.DAG;
+
+import java.util.Set;
+
+public interface JobInvocationObserver {
```
Comment (k-jamroz) [inline hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java:25]: you made this public API. Is that intended?

File: hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+    void onJobInvocation(long jobId, Set<MemberInfo> members, DAG dag, JobConfig jobConfig);
```
Comment (k-jamroz) [inline hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java:27]: when this observer is going to be invoked (add javadoc)? Should it be invoked in case of failures during job startup (currently it it invoked late)?

Comment (k-jamroz) [reply hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java:27]: it is also not invoked for normal jobs but you did not documented that fact. This is highly unexpected behavior and trap for us.

Comment (Fly-Style) [reply hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java:27]: Good catch, thanks.

File: hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java (hunk 6 above)
Comment (Fly-Style) [reply hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java:25]: Yes, I want this interface to be available everywhere.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/SqlEndToEndTestSupport.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        public JobConfig jobConfig;
+
+        @Override
+        public void onJobInvocation(DAG dag, JobConfig config) {
+            this.dag = dag;
+            this.jobConfig = config;
+        }
+    }
+
+    protected static class JobInvocationObserverImpl implements JobInvocationObserver {
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/SqlEndToEndTestSupport.java:137]: can't we use mockito mock+verify for the observers? wouldn't it be simpler and more flexible?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/misc/Pojo.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.misc;
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/misc/Pojo.java:17]: `misc` package contains tests to which this class is completely unrelated. Can you find a better place for it? We have a few `model` packages in tests.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/SqlEndToEndTestSupport.java (hunk 8 above)
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/SqlEndToEndTestSupport.java:137]: > wouldn't it be simpler and more flexible?

You're reviewing the simpler way for me now :)

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java (hunk 5 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java]: Done.

File: hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java (hunk 7 above)
Comment (Fly-Style) [reply hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java]: Done.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/misc/Pojo.java (hunk 9 above)
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/misc/Pojo.java]: Done.

**Review by ivanthescientist — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java (hunk 6 above)
Comment (k-jamroz) [reply hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java:33]: if so, this requires much more thinking (eg. do we want method specific to light jobs or general) and will have to keep backward compatibility. Public API also has to have @since and full javadoc (also for methods). Currently you documented public interface by using private implementation details - this will not be understandable. You also did not document mutability of paramters and blocking/cooperativeness requirements for the listener.

We can keep it as public API but annotate it as `@Beta`. I am almost sure that someone will (mis)use this API to implement some kind of Jet job monitoring system.

Comment (k-jamroz) [reply hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java:33]: BTW, observer is public API, but method to register it is private - it does not make sense.
I changed my mind: move the observer to `impl` package. If we really need to make it public, we always can do it.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/misc/Pojo.java (hunk 9 above)
Comment (k-jamroz) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/misc/Pojo.java]: still not a good place. `com.hazelcast.jet.sql.impl.schema` contains tests related to SQL schemas, `Pojo` is not used in them.

**Review by k-jamroz — APPROVED** [review]

**Comment by Fly-Style** [comment]
run-lab-run
